spm-monitor-utils/src/main/java/com/sematext/spm/client/status/AgentStatusRecorder.java

```diff
@@ -0,0 +1,190 @@
+    long now = System.currentTimeMillis();
+    
+    if (previousStatus == ConnectionStatus.CONNECTING || previousStatus == ConnectionStatus.FAILED) {
+      // we can update in any case
+      statusValues.put(StatusField.CONNECTION_STATUS, newStatus);  
+    } else {
+      // if connection was OK before, but now not anymore, we should wait for CONNECTION_OK_EXPIRY_TIME_MS before
+      // setting to failed (e.g. in case of json some URLs may be ok, some not because of new version of monitored
+      // service)
+      if (newStatus == ConnectionStatus.FAILED) {
+        if ((now - CONNECTION_OK_EXPIRY_TIME_MS) > lastConnectionOkStatusTime) {
+          statusValues.put(StatusField.CONNECTION_STATUS, newStatus);  
+        }
+      } else {
+        statusValues.put(StatusField.CONNECTION_STATUS, newStatus);
+      }
+    }
+    
+    if (newStatus == ConnectionStatus.OK) {
+      lastConnectionOkStatusTime = System.currentTimeMillis();
+    }
+
+    statusValues.put(StatusField.LAST_UPDATE, now);
+    
+    record();
+  }
+
+  public void updateConnectionStatus(ConnectionStatus newStatus, String newError) {
+    Map<String, Long> connErrors = (Map<String, Long>) statusValues.get(StatusField.CONNECTION_ERRORS);
+    connErrors.put(newError, System.currentTimeMillis());
+    updateConnectionStatus(newStatus);
+  }
+
+  public void updateMetricsCollected(boolean metricsCollected) {
+    statusValues.put(StatusField.METRICS_COLLECTED, metricsCollected);
+    statusValues.put(StatusField.LAST_UPDATE, System.currentTimeMillis());
+    record();
+  }
+
+  public void updateMetricsSent(boolean metricsSent) {
+    statusValues.put(StatusField.METRICS_SENT, metricsSent);
+    statusValues.put(StatusField.LAST_UPDATE, System.currentTimeMillis());
+    record();
+  }
+
+  public void addConnectionError(String errorMessage) {
+    Map<String, Long> connErrors = (Map<String, Long>) statusValues.get(StatusField.CONNECTION_ERRORS);
+    connErrors.put(errorMessage, System.currentTimeMillis());
+
+    statusValues.put(StatusField.LAST_UPDATE, System.currentTimeMillis());
+    
+    record();
+  }
+  
+  private void record() {
+    StringBuffer content = new StringBuffer(100);
+    for (StatusField key : statusValues.keySet()) {
+      content.append(key.getDisplayName());
+      content.append("=");
+      // raw serialization for now, but consider serializing e.g. errors differently
+      content.append(statusValues.get(key));
+      content.append(MonitorUtil.LINE_SEPARATOR);
+    }
+    
+    try {
+      FileUtil.write(content.toString(), statusFile);
+    } catch (IOException e) {
+      LOG.error("Can't write to status file", e);
+    }
+  }
+
+  private File getStatusFile() {
+    return new File(MonitorUtil.MONITOR_STATUS_FILE_DIR, getStatusFileName());
+  }
+
+  private String getStatusFileName() {
+    return appToken + "-" + subType + "-" + jvmName + "-monitor-" + processOrdinal + ".status";
```

Comment:
It might be tricky for STA to figure out the file name if the `processOrdinal` attribute changes over time.

Reply:
Here is why it exists - it is possible for user to start multiple agent instances that use the same spm-monitor-conf file. In that case these agents differ by the ordinal number. This would happen only in javaagent setups, so maybe STA could do this:
- check which lock file is used by particular service that has our javaagent running
- use the name of that lock file and just change suffix from .lock to .status

Doable?